### PR TITLE
Feature 12235 environment sample create edit

### LINKED ISSUE #12235: Add an edit form for environment samples in web app [5]
### Feature Description

#11721 has introduced a new environment sample entity. Create and edit forms are needed for users to be able to create new environment samples in the web app.

### Added Value/Benefit

Users can create environment samples required for one health surveillance.

### Acceptance Criteria

- [x] An edit form (also used for creation) for environment samples has been added with the following fields and headings:
  - Heading: "Laboratory sample"
    - UUID (read only; caption: Sample ID)
    - Environment name (read only)
    - Read-only label with the report date and reporting user (as already implemented for samples)
    - Date of sampling (required)
    - Sample material (required)
    - Specify sample material (if sample material == other, required in this case)
    - Field sample ID
  - Heading: Sample measurements
    - Volume (in mL)
    - Turbidity (in NTU)
    - Temperature (in °C) - pre-fill with values from -20 to +100 °C in steps of 0.1 (see symptoms for reference)
    - Total chlorine residue (mg/L)
    - pH value 
    - Weather condition at time of sampling - display each possible enum value as an individual checkbox
    - Heavy rain in the last 24 hours before sampling?
  - Heading: "Location of sampling site"
    - Location form (region, district, gps lat and gps lon required; location of the environment for which the sample is created is copied over and form is disabled by default)
    - Checkbox "Location differs from environment location" - if checked, clear and enable the location form; if unchecked, disable it and fill it with the environment location again; show this checkbox ABOVE the location form
  - Heading: "Sample management"
    - Laboratory - required
    - Laboratory details (if generic laboratory has been selected)
    - Subheading: "Requested pathogens to be tested"
    - Checkboxes for each pathogen in the system, similar to the requested pathogen tests for samples
    - Other requested pathogens to be tested
    - Checkbox "Sent/dispatched"
    - Dispatch date and dispatch details only if checkbox is checked
    - Checkbox "Received"
    - Receival date and lab sample ID only if checkbox is checked
    - Specimen condition
    - General comment
- [x] Environment sample entries in the directory can be opened and bring up the edit form
- [x] A button to create new environment samples has been added to the directory
- [x] Ensure that creating an environment sample with the location of an environment does not actually use the same location entity, but just copies over the values
- [x] Ensure that the form is properly read-only for users without the necessary jurisdiction and who are not the reporting users
- [x] Users with the `ENVIRONMENT_SAMPLE_EDIT_DISPATCH` user right can edit everything up until the "Received" checkbox; users with the `ENVIRONMENT_SAMPLE_EDIT_RECEIVAL` user right can only edit fields starting with the "Received" checkbox; The general comment field can be edited by everyone; users with both rights can edit everything
- [x] For creation forms, all users can edit everything up until the "Received" checkbox, but only users with the EDIT_RECEIVAL right can edit the receival information

### Implementation Details

_No response_

### Mockups

env_sample

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by leventegal-she**
Fixes #12235

### CODE REVIEW

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12533

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12533

File: sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreFacadeEjb.java [hunk 1]
```diff
@@ -197,24 +151,7 @@ protected String getDeleteReferenceField(DeletionReference deletionReference) {
...
-
-		cq.select(from.get(getDeleteReferenceField(entityConfig.getDeletionReference())));
-		cq.where(cb.equal(from.get(AbstractDomainObject.UUID), uuid));
-
-		Object result = em.createQuery(cq).getSingleResult();
-		return (Date) result;
-	}
-
-	protected abstract CoreEntityType getCoreEntityType();
+	protected abstract DeletableEntityType getCoreEntityType();
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreFacadeEjb.java:154]:
```suggestion
	protected abstract DeletableEntityType getDeletableEntityType();
```

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentSampleListComponent.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+import de.symeda.sormas.ui.utils.DateFormatHelper;
+import de.symeda.sormas.ui.utils.PaginationList;
+import de.symeda.sormas.ui.utils.components.sidecomponent.SideComponent;
+import de.symeda.sormas.ui.utils.components.sidecomponent.SideComponentField;
+
+public class EnvironmentSampleListComponent extends SideComponent {
+
+	private static final long serialVersionUID = -4946315430355463915L;
+
+	private final EnvironmentSampleList eventList;
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentSampleListComponent.java:50]:
```suggestion
	private final EnvironmentSampleList environmentSampleList;
```

File: sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentSampleListComponent.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+				Label resultLabel = new Label();
+				CssStyles.style(resultLabel, CssStyles.LABEL_BOLD, CssStyles.LABEL_UPPERCASE);
+
+				resultLabel.setValue(DataHelper.toStringNullable(sampleIndex.getSpecimenCondition()));
+				resultLabel.addStyleName(CssStyles.LABEL_WARNING);
+
+				topLeftLayout.addComponent(resultLabel);
+			}
+
+			Label dispatchRecieveLabel = new Label();
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/environment/EnvironmentSampleListComponent.java:156]:
```suggestion
			Label dispatchReceiveLabel = new Label();
```

File: sormas-ui/src/main/java/de/symeda/sormas/ui/samples/environmentsample/EnvironmentSampleController.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+import de.symeda.sormas.ui.utils.CommitDiscardWrapperComponent;
+import de.symeda.sormas.ui.utils.DateFormatHelper;
+import de.symeda.sormas.ui.utils.DeleteRestoreHandlers;
+import de.symeda.sormas.ui.utils.VaadinUiUtil;
+import de.symeda.sormas.ui.utils.components.automaticdeletion.DeletionLabel;
+import de.symeda.sormas.ui.utils.components.page.title.TitleLayout;
+
+public class EnvironmentSampleController {
+
+	public EnvironmentSampleController() {
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/samples/environmentsample/EnvironmentSampleController.java:55]: Isn't this redundant and can be removed?

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
Not the biggest fan of reproducing methods in EnvironmentSampleFacade that we already have in one of our interfaces, but I guess everything else would require a major refactoring, and that would definitely be out of scope for this issue, so it's fine.

Apart from that:

- otherSampleMaterial needs to be required if "Other" is selected
- Celsius symbol in temperature caption is not displayed correctly
- Fields that require a number should have a proper validation text if the user enters non-numeric characters (currently the text is "Could not convert value to Float"); see fields in AdditionalTestDto for examples
- Please remove the "Other" requested pathogen test from the checkbox list since we cover this by the additional text field; see my mobile app PR, I've done it there as well - the code should not fail if no pathogen named "OTHER" exists in the database
- Add some margin below the "Sent/Dispatched" and "Received" checkboxes
- As a National User/Admin WITHOUT the user right to edit receival information, I can still edit the receival information of environment samples
- As a National User/Admin WITHOUT the user right to edit dispatch information, I can still edit the dispatch information of environment samples
- As an Environment Officer, trying to open an environment sample through the side component of an environment sends me to a "The view could not be found" page

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12533

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
As an environment surveillance officer, I can edit environment samples that are not in my jurisdiction and for which I'm not the reporting user.

Also, please replace the temperature field with a normal text input field that only accepts float numbers; right now, it's not possible to put in floating point numbers, which would very likely result in accuracy issues, and since the field is a normal text input field in the mobile app, this would also result in display issues when a floating point number would be entered into the mobile app. Having a dropdown with all possible values is not really feasible in this situation I feel like, because we have a higher range of possible values than for human temperatures (which would result in A LOT of numbers in the dropdown list).

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
Temperature field has to accept floating point numbers

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12533

**Pull request merged**